adapters/undertone/undertone.go

```diff
@@ -0,0 +1,193 @@
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	imps, publisherId, errs := getImpsAndPublisherId(request)
+	if len(imps) == 0 {
+		return nil, errs
+	}
+
+	reqCopy := *request
+	reqCopy.Imp = imps
+
+	populateSiteApp(&reqCopy, publisherId, request.Site, request.App)
+	e := populateBidReqExt(&reqCopy)
+	if e != nil {
+		errs = append(errs, e)
+	}
+
+	requestJSON, err := json.Marshal(reqCopy)
+	if err != nil {
+		return nil, append(errs, err)
+	}
+
+	requestData := &adapters.RequestData{
+		Method: "POST",
+		Uri:    a.endpoint,
+		Body:   requestJSON,
+	}
+
+	return []*adapters.RequestData{requestData}, errs
+}
+
+func (a *adapter) MakeBids(request *openrtb2.BidRequest, requestData *adapters.RequestData, responseData *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if responseData.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if responseData.StatusCode == http.StatusBadRequest {
+		err := &errortypes.BadInput{
+			Message: "Unexpected status code: 400. Bad request from publisher. Run with request.debug = 1 for more info.",
+		}
+		return nil, []error{err}
+	}
+
+	if responseData.StatusCode != http.StatusOK {
+		err := &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info.", responseData.StatusCode),
+		}
+		return nil, []error{err}
+	}
+
+	var response openrtb2.BidResponse
+	if err := json.Unmarshal(responseData.Body, &response); err != nil {
+		return nil, []error{err}
+	}
+
+	impIdBidTypeMap := map[string]openrtb_ext.BidType{}
+	for _, imp := range request.Imp {
+		if imp.Banner != nil {
+			impIdBidTypeMap[imp.ID] = openrtb_ext.BidTypeBanner
+		} else if imp.Video != nil {
+			impIdBidTypeMap[imp.ID] = openrtb_ext.BidTypeVideo
+		}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(request.Imp))
+	bidResponse.Currency = response.Cur
+	for _, seatBid := range response.SeatBid {
+		for i, bid := range seatBid.Bid {
+			bidType, ok := impIdBidTypeMap[bid.ImpID]
+			if !ok {
+				continue
+			}
+
+			b := &adapters.TypedBid{
+				Bid:     &seatBid.Bid[i],
+				BidType: bidType,
+			}
+			bidResponse.Bids = append(bidResponse.Bids, b)
+		}
+	}
+	return bidResponse, nil
+}
+
+func populateBidReqExt(bidRequest *openrtb2.BidRequest) error {
+	undertoneBidderParams := &bidderParams{
+		Id:      adapterId,
+		Version: adapterVersion,
+	}
+	undertoneBidderParamsJSON, err := json.Marshal(undertoneBidderParams)
+	if err == nil {
+		extRequestPrebid := &openrtb_ext.ExtRequestPrebid{BidderParams: undertoneBidderParamsJSON}
+		bidRequestExt := &BidRequestExt{Prebid: extRequestPrebid}
+		bidRequestExtJSON, err2 := json.Marshal(bidRequestExt)
+		if err2 == nil {
+			bidRequest.Ext = bidRequestExtJSON
+		}
+	}
+	return err
+}
```

Comment:
missing error handling for `err2` on line 128 - what if json marshalling fails here.